File: sync.php
<?php

//@author Stanislav Polaniev <spolanyev@gmail.com>

declare(strict_types=1);

$startTime = microtime(true);
$wordCount = 0;
$selectedWords = [];

$fullPathDirectory = realpath(dirname(__DIR__) . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR . 'a');
if (is_dir($fullPathDirectory)) {
    if (false !== ($directory = dir($fullPathDirectory))) {
        while (false !== ($entry = $directory->read())) {
            if ('.' !== $entry && '..' !== $entry) {
                $wordCount++;
                $fullPathFrequency = $fullPathDirectory . DIRECTORY_SEPARATOR . $entry . DIRECTORY_SEPARATOR . '2015-2017-spoken-frequency.txt';
                if (is_file($fullPathFrequency)) {
                    $number = file_get_contents($fullPathFrequency);
                    if (intval($number) >= 798) {
                        $fullPathTranslation = $fullPathDirectory . DIRECTORY_SEPARATOR . $entry . DIRECTORY_SEPARATOR . 'translation.txt';
                        if (is_file($fullPathTranslation)) {
                            $word = '';
                            if ($filePointer = fopen($fullPathTranslation, 'r')) {
                                $line = fgets($filePointer, 128);
                                fclose($filePointer);
                                if (false !== $line) {
                                    $word = trim($line);
                                }
                            }

                            if ('' === $word) {
                                $word = $entry;
                            }
                            $selectedWords[] = $word;
                        }
                    }
                }
            }
        }
        $directory->close();
    }
}

echo sprintf(
    "selected %d words from %d, took %.3f seconds\n",
    count($selectedWords),
    $wordCount,
    round(microtime(true) - $startTime, 3)
);
/*
foreach ($selectedWords as $i => $word) {
    echo sprintf("%d. %s\n", ++$i, $word);
}
*/

File: sync.cpp
//@author Stanislav Polaniev <spolanyev@gmail.com>

#include <chrono>
#include <iostream>
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>
#include <ranges>
#include <format>

int main() {
    auto start_time = std::chrono::high_resolution_clock::now();
    int word_count = 0;
    std::vector<std::string> selected_words;

    std::filesystem::path full_path_directory = std::filesystem::canonical(
            "./../../../private/speed-comparison/data/a");
    if (std::filesystem::exists(full_path_directory)) {
        for (const auto entry: std::filesystem::directory_iterator(full_path_directory)
                               | std::views::transform([](const auto &entry) {
            return std::make_pair(entry, entry.path() / "2015-2017-spoken-frequency.txt");
        })) {
            word_count++;
            auto &[entry_info, full_path_frequency] = entry;
            if (std::filesystem::exists(full_path_frequency)) {
                if (std::ifstream frequency(full_path_frequency); frequency) {
                    int number;
                    frequency >> number;
                    if (number >= 798) {
                        std::filesystem::path full_path_translation = entry_info.path() / "translation.txt";
                        if (std::filesystem::exists(full_path_translation)) {
                            std::string word;
                            if (std::ifstream translation{full_path_translation}; std::getline(translation, word)) {
                                //trim
                                word.erase(0, word.find_first_not_of(" \t\n\r"));
                                word.erase(word.find_last_not_of(" \t\n\r") + 1);
                            }

                            if (word.empty()) {
                                word = entry_info.path().filename().string();
                            }
                            selected_words.push_back(word);
                        }
                    }
                }
            }
        }
    }

    std::cout << std::format("selected {} words from {}, took {:.3f} seconds\n", selected_words.size(), word_count,
                             std::chrono::duration<double>(
                                     std::chrono::high_resolution_clock::now() - start_time).count());
    /*
    size_t i = 0;
    for (const auto &word: selected_words) {
        std::cout << std::format("{}. {}\n", ++i, word);
    }
    */
    return 0;
}
